parser/impl/parser.cpp

```diff
@@ -7,51 +7,289 @@ namespace parser
+                m_lexer.Next();
+            }
+            return nullptr;
+        }
+        parameterNode.type = parseType();
+        return std::make_shared<ast::Parameter>(parameterNode);
+    }
+
+    sPtr<ast::Type> Parser::parseType() {
+        lexer::Token currentToken = m_lexer.Peek();
+        if (lexer::TokenType *type = std::find(std::begin(primitives), std::end(primitives), currentToken.type);
+            type != std::end(primitives)) {
+            ast::PrimitiveType typeNode;
+            typeNode.type = currentToken;
+            m_lexer.Next();
+            return std::make_shared<ast::PrimitiveType>(typeNode);
+        } else if (currentToken.type == lexer::TokenType::Array) {
+            return parseArrayType();
+        } else if (currentToken.type == lexer::TokenType::Record) {
+            return parseRecordType();
+        } else if (currentToken.type == lexer::TokenType::Identifier) {
+            ast::AliasedType typeNode;
+            typeNode.name = currentToken;
+            m_lexer.Next();
+            return std::make_shared<ast::AliasedType>(typeNode);
+        } else {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Unknown type",
+            });
+            m_lexer.Next();
+            return nullptr;
+        }
+    }
+
+    sPtr<ast::ArrayType> Parser::parseArrayType() {
+        lexer::Token currentToken = m_lexer.Next();
+        if (currentToken.type != lexer::TokenType::Array) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Expected 'array' keyword but did not find it"
+            });
+            return nullptr;
+        }
+        ast::ArrayType arrayNode;
+        if (m_lexer.Peek().type == lexer::TokenType::OpenBrack) {
+            m_lexer.Next();
+            arrayNode.length = parseExpression();
+            currentToken = m_lexer.Next();
+            if (currentToken.type != lexer::TokenType::CloseBrack) {
+                m_errors.push_back(Error{
+                    .pos = currentToken.pos,
+                    .message = "Unexpected token"
+                });
+                while(m_lexer.Next().type != lexer::TokenType::CloseBrack);
+                return nullptr;
+            }
+        }
+        arrayNode.elementType = parseType();
+        return std::make_shared<ast::ArrayType>(arrayNode);
+    }
+
+    sPtr<ast::RecordType> Parser::parseRecordType() {
+        lexer::Token currentToken = m_lexer.Next();
+        if (currentToken.type != lexer::TokenType::Record) {
+            m_errors.push_back(Error{
+                .pos = currentToken.pos,
+                .message = "Expected 'record' keyword but did not find it"
+            });
+            return nullptr;
+        }
+        ast::RecordType recordNode;
+        while (m_lexer.Peek().type == lexer::TokenType::NewLine) m_lexer.Next();
```

Comment:
Why not `skipWhile(isNewLine)`?

Reply:
Because it consumes the next token after the new line (and returns it), but I want the next token to be consumed by the call to `parseVariable`